### ISSUE TITLE
Target 'install-strip' breaks unilinear chain of rules

### ISSUE THREAD

**Issue by katjav** [pure-data/pd-lib-builder#45]
Target `install-strip` was introduced as a response to extremely bloated executables from Mingw-w64, as an outcome of discussion in issue #42. Size reduction after stripping doesn't normally outweigh the reduction of options for troubleshooting, but that is different with Mingw-w64 where non-stripped sizes are up to 10 times larger, no matter if built with static standard lib or not.

With target `install-strip` as an alternative to regular `install` which doesn't strip, stripping appears to be optional, a choice up to anyone who builds. But that is no longer the case when for example:

1. a library makefile defines `install` as prerequisite of a custom target, or
2. a library makefile calls `make install` as a submake process in a custom target, or
3. a central makefile in a multi-lib project calls `make install`

Think of test targets or special packaging targets in a lib makefile. Considering such use cases, 'parallel' installation targets aren't a good solution to provide optional stripping. In general a chain of `make` rules should be unilinear, but `install-strip` formally and effectively introduces a diverging build route rather than an extra option on a given route. This is not justified since stripped and non-stripped executables are supposed to be identical regarding their functionality.

A variable instead of alternative target could keep the chain of rules in pdlibbuilder unilinear from source files to installed executables, so lib makefiles and project makefiles can rely on that. Stripping could be done like:

`make install strip=yes`

with a strip 'function' (multiline define) which does nothing by default and only strips when strip=yes, in target `install-executables`.

**Comment by katjav** [comment]
To illustrate the chain of rules without and with target `install-strip`:

`source files < object files < executables < all < install`

`source files < object files < executables < all < install < install-strip`

The arrow `<` indicates what is dependent on what. The chain of rules guarantees that if you install a libdir it is verified that the executables are built and also that they are up to date with the latest changes in source files. As a side effect you can do the whole chain from compilation to installation with a single command `make install` or `make install-strip` respectively, though it isn't advertised.

Now imagine your lib makefile adds custom targets to the chain, depending on proper installation. Like a test program, a distribution target, or anything. And now pdlibbuilder comes with _two_ installation targets. You can't make both prerequisite. This makes no sense:

`test: install install-strip`

I would be pissed.

**Comment by katjav** [comment]
Figured out an implementation where stripping is done conditionally based on a user variable. I've tested that the method is effective with recursive `make install` and with targets that have `install` as prerequisite (tested Mingw-w64 cross compile, Linux and OSX). It works straightforward in all sorts of situations and I'm happy with it. But there's one question still, about the variable name. I had suggested: 

`make install strip=yes`

While this looks straightforward as a command line, it is confusing when considering the name of the stripping program itself (strip) and the makefile variable for that program (STRIP), plus the fact that `make` itself has a function (strip) that strips spaces from `make` variables. In particular the specification of pdlibbuilder variables would look akward with:

```
STRIP: name of the strip program (default: strip)
strip: option for target 'install' ('strip=yes' to strip executables; no default definition)
```
A more specific variable name is needed for the stripping option. The best I can think of is `dostrip` to indicate an action. Something like `stripexecutables` or `stripsymbols` would be more specific but annoyingly long for command line typing. So `dostrip` it will be if nothing better comes up shortly, then I'll push yet another branch for test and review.

**Comment by Lucarda** [comment]
Cool.
Yesterday I contributed to porres/cyclone w32/64 builds. Stripping was a "MUST" on Msys2.
: )

**Comment by umlaeute** [comment]
since the `.dll` files are hopefully not executable, i think that `stripbinaries` is more appropriate.

**Comment by katjav** [comment]
Commit 72f91f5 on branch `strip-variable` implements above-mentioned method. Use like:

`make install dostrip=yes`

**Comment by Lucarda** [comment]
72f91f5 is working nice on Msys1 & Msys2 32/64.

Note: Is not an issue for me but the later `install-strip` give 2KB less per file. We can ignore this.

**Comment by katjav** [comment]
Happy to hear that it works. The 2KB difference is because `--strip-unneeded` is used which leaves a few symbols. To me this sounded like the safest option, but now I'm reading that `--strip-unneeded` is only required for library archives which is irrelevant in our case:

https://www.technovelty.org/linux/stripping-shared-libraries.html

So I'll do away with `--strip-unneeded`. For some reason OSX requires strip option `-x` to keep global symbols. Although not required for Linux I'll use the same option here. For Windows all symbols will be stripped because `-x` makes them still too big. Updated in commit 66ad3ce.

**Comment by katjav** [comment]
> since the `.dll` files are hopefully not executable, i think that `stripbinaries` is more appropriate.

True, I'm always wrongly speaking of executables when referring to dlls, it's throughout pdlibbuilder.

**Comment by Lucarda** [comment]
All good with 66ad3ce.

Tested with Msys1 & 2 (32/64).

Also tested the resulting binaries and they are 2KB smaller.

: )

**Comment by katjav** [comment]
So it seems that this stripping method works satisfactorily for all platforms now. One thing still worries me though: users may forget to use the strip option, or more likely, be unaware of its existence. Considering the outrageous size of Mingw-w64 builds I'd rather like to strip these by (overridable) default.

For reference: Pd-extended's Makefile Template strips installed binaries by default for all platforms. Externals (bob~ & Co.) in Pd-0.48-1.msw come with symbols stripped. I haven't built Pd for Windows myself and don't know if Miller's distribution was stripped 'by hand'? For Linux the build system doesn't strip by default.

**Comment by Lucarda** [comment]
> Considering the outrageous size of Mingw-w64 builds I'd rather like to strip these by (overridable) default.


Sure. Why not?
You mean _strip by default_  when using `install`?

**Comment by Lucarda** [comment]
Sounds good to me. :)

**Comment by katjav** [comment]
Above mentioned suggestion is implemented with commit fb84ccd. With this, stripping is default (but overridable) behavior for the Windows case and an option for other systems. Tested with Linux and Mingw-w64 crosscompiler builds. 

Since the proposal of default stripping for Windows didn't attract objections it seems that this can be implemented on the master branch after verification. @Lucarda, does it work on your systems?

**Comment by Lucarda** [comment]
fb84ccd working nice here.

Msys1 (32) Msys2 (32 & 64).

`install` _strips_ by default and it can be overrided with `stripflags=`

: )

**Comment by katjav** [comment]
@Lucarda thanks for all your patient testing. Branch 'strip-variable' with all its experiments isn't merged as such, the end result is now implemented in master with commit 397b72e.

**Comment by katjav** [comment]
To summarize the result of this issue, target 'install-strip' is replaced by a method where:

- the value of variable 'stripflags' determines if and how stripping will be done after install
- if variable 'stripflags' is not defined or defined empty, no stripping will be done
- variable 'stripflags' is defined '--strip-all' for Windows by default, and not defined for other systems
